composer/core/state.py

```diff
@@ -315,36 +334,33 @@ def load_state_dict(self, state: types.StateDict, strict: bool = False):
+        state = _ensure_backwards_compatible_checkpointing(state)
+
+        for attribute_name, serialized_value in state.items():
+            if attribute_name not in self.serialized_attributes:
+                # it's possible some attributes we removed
                 continue
-            serialized_value = state[state_field_name]
 
-            if state_field_name == "model":
+            if attribute_name == "model":
                 self.load_model_state(state, strict=strict)
+                continue
+            state_field_value = getattr(self, attribute_name)
+            if attribute_name in _STATE_DICT_SERIALIZED_ATTRIBUTES:
+                for target in ensure_tuple(state_field_value):
+                    if type(target).__qualname__ not in serialized_value:
+                        warnings.warn(
+                            f"{type(target).__qualname__} is not in the state_dict. Its state will not be restored.",
+                            category=UserWarning)
+                        continue
+                    source = serialized_value[type(target).__qualname__]
+                    target.load_state_dict(source)
             else:
-                # Duck typing since runtime checkable protocols are not available in Python 3.7
-                is_state_dict_serializable = any(hasattr(x, "load_state_dict") for x in ensure_tuple(state_field_value))
-                if len(ensure_tuple(state_field_value)) > 0:
-                    # any and all should be the same, except in for an empty collection, since
-                    # any([]) is False, while all([]) is True
-                    if is_state_dict_serializable != all(
-                            hasattr(x, "load_state_dict") for x in ensure_tuple(state_field_value)):
-                        raise RuntimeError(
-                            f"Every member of {state_field_name} should support `load_state_dict`, or no members should support it."
-                        )
-                if is_state_dict_serializable:
-                    for target in ensure_tuple(state_field_value):
-                        if target.__class__.__qualname__ not in serialized_value:
-                            warnings.warn(
-                                f"{target.__class__.__qualname__} was not found in the state_dict. Its state will NOT be restored",
-                                category=UserWarning)
-                            continue
-                        source = serialized_value[target.__class__.__qualname__]
-                        target.load_state_dict(source)
-                else:
-                    # direct serialization
-                    setattr(self, state_field_name, serialized_value)
+                # direct serialization
+                try:
+                    setattr(self, attribute_name, serialized_value)
+                except AttributeError:
+                    # ignore AttributeError for properties that have getters but not setters.
```

Comment:
Log a warning here if it's not something that should happen often?

Reply:
This is by design and will happen often -- `is_module_ddp` has no setter, and we'll hit this on every checkpoint load. Considering that the setattr is on `self` (and we don't define `__slots__`), this would only happen on properties that don't have a setter. Since the `attribute_name` comes from `self.serialized_attributes` (and is not something that the user sets, or is defined in the checkpoint), I think this should be safe?

Reply:
Ah okay sg